plutus-core/untyped-plutus-core/test/Transform/Simplify.hs

```diff
@@ -36,6 +36,75 @@ basicInline = runQuote $ do
     n <- freshName "a"
     pure $ Apply () (LamAbs () n (Var () n)) (mkConstant @Integer () 1)
 
+mkInlinePurityTest ::
+    Quote (Term Name PLC.DefaultUni PLC.DefaultFun ()) ->
+    Term Name PLC.DefaultUni PLC.DefaultFun ()
+mkInlinePurityTest termToInline = runQuote $ do
+    a <- freshName "a"
+    b <- freshName "b"
+    -- In `[(\a . \b . a) termToInline]`, `termToInline` will be inlined
+    -- if and only if it is pure.
+    Apply () (LamAbs () a $ LamAbs () b $ Var () a) <$> termToInline
+
+-- | A single @Var@ is pure.
+inlinePure1 :: Term Name PLC.DefaultUni PLC.DefaultFun ()
+inlinePure1 = mkInlinePurityTest $ Var () <$> freshName "a"
+
+-- | @force (delay a)@ is pure.
+--
+-- Note that this relies on @forceDelayCancel@ to cancel the @force@ and the @delay@,
+-- otherwise the inliner would treat the term as impure.
+inlinePure2 :: Term Name PLC.DefaultUni PLC.DefaultFun ()
+inlinePure2 = mkInlinePurityTest $ Force () . Delay () . Var () <$> freshName "a"
+
+-- | @[(\x -> \y -> [x x]) (con integer 1)]@ is pure.
+--
+-- Note that the @(con integer 1)@ won't get inlined: it isn't pre-inlined because
+-- @x@ occurs twice, and it isn't post-inlined because @costIsAcceptable Constant{} = False@.
+-- However, the entire term will be inlined since it is pure.
+inlinePure3 :: Term Name PLC.DefaultUni PLC.DefaultFun ()
+inlinePure3 = mkInlinePurityTest $ do
+    x <- freshName "x"
+    y <- freshName "y"
+    pure $
+        Apply
+            ()
+            (LamAbs () x $ LamAbs () y $ Apply () (Var () x) (Var () x))
+            (mkConstant @Integer () 1)
+
+-- | @force ([(\x -> delay (\y -> [x x])) (delay ([error (con integer 1)]))])@ is pure.
+--
+-- The @error@ is not evaluated when evaluating this term.
+inlinePure4 :: Term Name PLC.DefaultUni PLC.DefaultFun ()
+inlinePure4 = mkInlinePurityTest $ do
```

Comment:
I will say that I find it much harder to review these tests than ones written with literal PIR. It's much easier to look at the diffs between two PIR programs than between this bit of haskell code and the PIR program...

I made several erroneous comments that I had to delete.

(cc @effectfully )

Reply:
> I will say that I find it much harder to review these tests than ones written with literal PIR. It's much easier to look at the diffs between two PIR programs than between this bit of haskell code and the PIR program...

> I made several erroneous comments that I had to delete.

I think the solution is trivial here: print the unoptimized PIR code alongside the optimized one. NOT write PIR by hand. Only a very few extremely evil people deserve this kind of punishment.